File: compiler/testData/loadJava/compiledJava/kotlinSignature/propagation/parameter/DeeplySubstitutedClassParameter.java
package test;

public interface DeeplySubstitutedClassParameter {

    public interface Super<T> {
        void foo(T p);

        void dummy(); // to avoid loading as SAM interface
    }

    public interface Middle<E> extends Super<E> {
        void foo(E p);
    }

    public interface Sub extends Middle<String> {
        void foo(String p);
    }
}

File: kotlin-ide/idea/testData/compiler/loadJava/compiledKotlin/fromLoadJava/kotlinSignature/propagation/return/DeeplySubstitutedClassParameter.kt
package test

public interface DeeplySubstitutedClassParameter {

    public interface Super<T> {
        public fun foo(): T

        public fun dummy() // to avoid loading as SAM interface
    }

    public interface Middle<E>: Super<E> {
        override fun foo(): E
    }

    public interface Sub: Middle<String> {
        override fun foo(): String
    }
}
